Loki: Add parsing of query to visual query

ivanahuckova opened the pull request:
**What this PR does / why we need it**:
This PR adds parsing between text query and visual query. I

What is missing and could be part of the next PR: 
- [x] unwrap
- [x]  grouping (e.g sum_by), but this is currently not supported in visual query editor either
- [x]  implement line_filter in visual query editor
- [x]  implement label_filter in visual query editor


Part of: 

https://github.com/grafana/grafana/issues/44253

ivanahuckova committed:
Create parser

ivanahuckova committed:
Add parsing

ivanahuckova committed:
Update comment

ivanahuckova committed:
Remove operations that we don't support

ivanahuckova committed:
Merge remote-tracking branch 'origin' into ivana/loki-visual-parsing

ivanahuckova committed:
Resolve type errors

ivanahuckova committed:
Update test

torkelo commented:
If I use ticks for string they get added as ticks inside strings in the model 
![Screenshot from 2022-03-22 09-04-58](https://user-images.githubusercontent.com/10999/159440238-0e4e2d3a-fb35-479f-8672-900a90c30279.png)
![Screenshot from 2022-03-22 09-36-44](https://user-images.githubusercontent.com/10999/159440268-4dd00bcf-a368-4151-9cfc-7f5e5486c9b4.png)

ivanahuckova committed:
Handle backticks

ivanahuckova committed:
Handle backticks

ivanahuckova committed:
Merge remote-tracking branch 'origin' into ivana/loki-visual-parsing

ivanahuckova committed:
Remove copied test, update test

ivanahuckova commented:
> If I use ticks for string they get added as ticks inside strings in the model

Fixed. I have also added functionality that unescapes `\` if user uses `"` and is escaping it.

aocenas commented on public/app/plugins/datasource/loki/querybuilder/LokiQueryModeller.ts:
```diff
@@ -27,8 +27,13 @@ export class LokiQueryModeller extends LokiAndPromQueryModellerBase<LokiVisualQu
+      queryString = this.renderBinaryQueries(queryString, query.binaryQueries);
+    } catch (err) {
+      // TODO: For now we just log it
+      console.error(err);
```
oh ok but this is a different part. I thought we talked about the parsing. Here in rendering erroring out seems a bit weird. like you have an operation in the vis query object but at the same time, you don't know how to render it? In that case, how did it even get into that object? If it was added by the user I would assume we have a definition for it (and so also render code) or if it was parsed from string but we don't have a definition that could be seen as a parsing bug because we also don't know what UI to use for such operation.

aocenas commented on public/app/plugins/datasource/loki/querybuilder/components/LokiQueryEditorSelector.test.tsx:
```diff
@@ -125,29 +132,30 @@ describe('LokiQueryEditorSelector', () => {
-  //   expect(screen.getByText('$__interval')).toBeInTheDocument();
-  // });
+  it('parses query when changing to builder mode', async () => {
+    jest.spyOn(console, 'error').mockImplementation();
```
Why is this here?

aocenas replied:
Seems like there was unchecked async setState which triggered an error, so fixed that and this is not needed anymore

aocenas commented on public/app/plugins/datasource/loki/querybuilder/components/LokiQueryEditorSelector.tsx:
```diff
@@ -18,10 +19,24 @@ export const LokiQueryEditorSelector = React.memo<LokiQueryEditorProps>((props)
-      onChange({ ...query, editorMode: newMetricEditorMode });
+      const change = { ...query, editorMode: newMetricEditorMode };
+      if (newMetricEditorMode === QueryEditorMode.Builder) {
+        const result = buildVisualQueryFromString(query.expr);
```
For some other PR probably, but I think this (parsing + error handling modal) could be moved into a separate hook and reused.

aocenas committed:
Merge branch 'main' into ivana/loki-visual-parsing

aocenas committed:
Parsing for binary operations

aocenas committed:
Remove error about setting state after unmount
